### PR TITLE
Add TensorFlow estimator hook

### PR DESCRIPTION
**Pull request by convoliution**
This basically only logs loss.

Screen Shot 2020-01-21 at 2 21 20 PM
Screen Shot 2020-01-21 at 2 21 32 PM

### CODE REVIEW

File: verta/verta/integrations/tensorflow/__init__.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                _utils.validate_flat_key(value.tag)
+            except ValueError:  # key has slashes, probably (usually properties of specific layers)
+                continue
+
+            if value.WhichOneof("value") == "simple_value":
+                try:
+                    self.run.log_observation(value.tag, value.simple_value)
+                except:
+                    pass  # don't halt execution
+            # TODO: support other value types
```
Comment (convoliution) [inline verta/verta/integrations/tensorflow/__init__.py]: [Other value types](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/summary.proto#L112) are artifacts, so lower priority to support at the moment.

File: verta/verta/integrations/tensorflow/__init__.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+        summary_msg = _parse_summary_proto_str(run_values.results['summary'])
+
+        for value in summary_msg.value:
+            # TODO: something better than this check-skip
+            #     The Client currently only allows alphanum, dashes, and underscores in keys.
+            try:
+                _utils.validate_flat_key(value.tag)
+            except ValueError:  # key has slashes, probably (usually properties of specific layers)
+                continue
```
Comment (convoliution) [inline verta/verta/integrations/tensorflow/__init__.py]: As an example of a value tag with slashes:
```json
[
    {"tag": "linear/bias", "simple_value": 0.0},
    {"tag": "linear/fraction_of_zero_weights", "simple_value": 1.0}
]
```

I'd prefer to not make a wider Client change (allow slashes in keys) to log these names as-is. Maybe I could replace the slashes with other characters, but that feels like a hacky workaround that users still wouldn't appreciate.

File: verta/tests/test_integrations.py [hunk 3]
```diff
@@ -122,3 +122,38 @@ def test_functional_api(self, experiment_run):
...
+        label_series = pd.Series(np.random.randint(0, 2, size=samples))
+
+        feature_columns = []
+        for feature_name in data_df.columns:
+            feature_columns.append(tf.feature_column.numeric_column(feature_name, dtype=tf.float32))
+
+        def train_input_fn():
+            ds = tf.data.Dataset.from_tensor_slices((dict(data_df), label_series))
+            ds = ds.shuffle(1000)
+            ds = ds.batch(32).repeat(10)
```
Comment (conradoverta) [inline verta/tests/test_integrations.py]: Remove bc we don't need this in unit tests

File: verta/tests/test_integrations.py [hunk 4]
```diff
@@ -122,3 +122,38 @@ def test_functional_api(self, experiment_run):
...
+    def test_estimator_hook(self, experiment_run):
+        verta_integrations_tensorflow = pytest.importorskip("verta.integrations.tensorflow")
+        VertaHook = verta_integrations_tensorflow.VertaHook
+
+        np = pytest.importorskip("numpy")
+        pd = pytest.importorskip("pandas")
+        tf = pytest.importorskip("tensorflow")
+
+        # adapted from https://www.tensorflow.org/tutorials/estimator/linear
+        samples = 500
```
Comment (conradoverta) [inline verta/tests/test_integrations.py]: 5

File: verta/tests/test_integrations.py [hunk 5]
```diff
@@ -122,3 +122,38 @@ def test_functional_api(self, experiment_run):
...
+            feature_columns.append(tf.feature_column.numeric_column(feature_name, dtype=tf.float32))
+
+        def train_input_fn():
+            ds = tf.data.Dataset.from_tensor_slices((dict(data_df), label_series))
+            ds = ds.shuffle(1000)
+            ds = ds.batch(32).repeat(10)
+            return ds
+
+        linear_est = tf.estimator.LinearClassifier(feature_columns=feature_columns)
+        linear_est.train(train_input_fn, hooks=[VertaHook(experiment_run, every_n_steps=5)])
```
Comment (conradoverta) [inline verta/tests/test_integrations.py]: `every_n_steps=1`

File: verta/verta/integrations/tensorflow/__init__.py (hunk 2 above)
Comment (conradoverta) [reply verta/verta/integrations/tensorflow/__init__.py]: Why can't we have slash?

**Comment by convoliution** [comment]
@conradoverta 
>Need to add support for TF 2.0 too if that's different.
This works in TF 2.0 as-is.

File: verta/verta/integrations/tensorflow/__init__.py (hunk 2 above)
Comment (convoliution) [reply verta/verta/integrations/tensorflow/__init__.py]: No particular technical reason.
Client has a whitelist for characters to use in keys, in attempt to future-proof it.
Thinking about it on the drive home, slashes shouldn't be a problem.

**Comment by conradoverta** [comment]
> @conradoverta
> 
> > Need to add support for TF 2.0 too if that's different.
> > This works in TF 2.0 as-is.

Are you sure? `tensorflow.compat.v1` indicates to me that it's `v1` specific, as iirc, the things that work the same between both are not in that package.

**Comment by convoliution** [comment]
@conradoverta I have been running this test on my machine with TensorFlow 2.0 and 2.1, and it succeeds.

My understanding has also been that `tensorflow.compat.v1` is specifically provided so users in 2.X can _continue_ using the 1.X API, at the cost of not being able to leverage 2.X's improvements.
I can make the change, but that currently means sinking time into figuring out where the `Session` protobuf object went in 2.0 (I know where `SessionRunArgs` and `SessionRunHook` went).

**Comment by conradoverta** [comment]
I'm not sure everything inside `tensorflow.compat.v1` is compatible with the TF 2.0 API. At least that was not the original proposal in the beginning. I might be wrong, but we should double-check.

This is not "sinking" time. It's about doing it properly. I'd say it's very reasonable to assume that "supporting TF" means supporting the common use cases at least, which includes both APIs currently. I agree this PR should cover people with TF 2.0 installed in their system that use TF 1.X API. I'm not sure it covers people using the TF 2.0 API. We can merge this in as TF 1.X API support.

**Comment by convoliution** [comment]
Okay. I was confident in my understanding of my use of `compat.v1` so I thought it was already "done properly", but if I'm wrong then it's certainly worth the extra time to change it.

**Comment by convoliution** [comment]
I wanted to tackle the 2.X support this morning, to avoid revisiting it down the line and losing context.

The `Summary` protobuf, `SessionRunArgs`, and `SessionRunHook` are now all imported directly from their proper places.

The only thing unaccounted for is `tensorflow.summary.merge_all()`, which was [removed altogether in 2.X](https://www.tensorflow.org/tensorboard/migrate); even the most relevant built-in Estimator Hook [suggests that this functionality is tied directly to 1.X](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/basic_session_run_hooks.py#L798). A similar Hook provided by a different Client library still uses `tensorflow.summary.merge_all()`, suggesting to me that full 2.X support for Estimators may be beyond my individual capabilities.

**Pull request merged**